components/net/image_cache.rs

```diff
@@ -433,33 +487,133 @@ struct ImageCacheStore {
     /// Cross-process compositor API instance.
     #[ignore_malloc_size_of = "Channel from another crate"]
     compositor_api: CrossProcessCompositorApi,
+
+    pipeline_id: Option<PipelineId>,
+
+    /// Main struct to handle the cache of `WebRenderImageKey` and
+    /// images that do not have a key yet.
+    key_cache: KeyCache,
 }
 
 impl ImageCacheStore {
-    // Change state of a url from pending -> loaded.
-    fn complete_load(&mut self, key: LoadKey, mut load_result: LoadResult) {
+    /// Finishes loading the image by setting the WebRenderImageKey and calling `compete_load` or `complete_load_svg`.
+    fn set_key_and_finish_load(&mut self, pending_image: PendingKey, image_key: WebRenderImageKey) {
+        match pending_image {
+            PendingKey::RasterImage((pending_id, mut raster_image)) => {
+                set_webrender_image_key(&self.compositor_api, &mut raster_image, image_key);
+                self.complete_load(pending_id, LoadResult::LoadedRasterImage(raster_image));
+            },
+            PendingKey::Svg((pending_id, mut raster_image, requested_size)) => {
+                set_webrender_image_key(&self.compositor_api, &mut raster_image, image_key);
+                self.complete_load_svg(raster_image, pending_id, requested_size);
+            },
+        }
+    }
+
+    /// Load an image. This will either load the image normally, or wait till a key
+    /// is available and then load the image. Only call this if the image does not have a `LoadKey` yet.
+    fn load_image_with_keycache(&mut self, pending_image: PendingKey) {
+        if let Some(pipeline_id) = self.pipeline_id {
+            match self.key_cache.cache {
+                KeyCacheState::PendingBatch => {
+                    self.key_cache.images_pending_keys.push_back(pending_image);
+                },
+                KeyCacheState::Ready(ref mut cache) => match cache.pop_front() {
+                    Some(image_key) => {
+                        self.set_key_and_finish_load(pending_image, image_key);
+                        self.key_cache.cache = KeyCacheState::PendingBatch;
```

Comment:
This line should be removed; the state should remain `Ready`, and only change when empty, after having requested a new batch.